File: ThreeJS/src/game/pinball-knight/map-overlay.ts
/**
 * THE FLOOR MAP — now a delegate.
 *
 * The painting was always in `map-render.ts`; this module was the chrome around
 * it (a fixed div, a canvas, a resize listener, a RAF). All of that is gone —
 * `gui/screens/floor-map.ts` gives the same painter a rect inside the pixel
 * pass. The suppression flag stays here because it is a RULE about when the map
 * is available (the site map yields the key for the run), not presentation.
 */
import { floorMapScreen } from "./gui/screens/floor-map";
import { close as closeUiScreen, isOpen as uiIsOpen, push as pushUiScreen } from "./gui/stack";

let suppressed = false;

/** The site map yields the map key for the run; this is that gate. */
export function setMapSuppressed(on: boolean): void {
  suppressed = on;
  if (on) closeUiScreen("floor-map");
}

export function isFloorMapOpen(): boolean {
  return uiIsOpen("floor-map");
}

export function closeFloorMap(): void {
  closeUiScreen("floor-map");
}

/** Returns whether the map is open AFTER the toggle. */
export function toggleFloorMap(_container?: HTMLElement): boolean {
  if (suppressed) return false;
  if (uiIsOpen("floor-map")) {
    closeUiScreen("floor-map");
    return false;
  }
  pushUiScreen(floorMapScreen());
  return true;
}

File: Rust/crates/pk-gui/src/map_overlay.rs
//! Floor Map Overlay State Gate — Modal availability gating and screen toggle state.
//!
//! PORTS: `map-overlay.ts`

#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub struct MapOverlayState {
    pub suppressed: bool,
    pub is_open: bool,
}

impl MapOverlayState {
    pub fn new() -> Self {
        Self::default()
    }

    /// Sets suppression state. When suppressed, closes the map if open.
    pub fn set_map_suppressed(&mut self, on: bool) {
        self.suppressed = on;
        if on {
            self.is_open = false;
        }
    }

    pub fn is_floor_map_open(&self) -> bool {
        self.is_open
    }

    pub fn close_floor_map(&mut self) {
        self.is_open = false;
    }

    /// Toggles the map overlay and returns whether the map is open AFTER the toggle.
    pub fn toggle_floor_map(&mut self) -> bool {
        if self.suppressed {
            return false;
        }
        if self.is_open {
            self.is_open = false;
            false
        } else {
            self.is_open = true;
            true
        }
    }
}
